include/AdePT/core/PerEventScoringImpl.cuh

```diff
@@ -0,0 +1,347 @@
+
+      auto bufferBegin = buffer.hitScoringInfo.hitBuffer_dev;
+
+      cub::DeviceMergeSort::SortKeys(fGPUSortAuxMemory.get(), fGPUSortAuxMemorySize, bufferBegin,
+                                    buffer.hitScoringInfo.fSlotCounter, CompareGPUHits{}, cudaStreamForHitCopy);
+
+      COPCORE_CUDA_CHECK(cudaMemcpyAsync(buffer.hostBuffer, bufferBegin,
+                                        sizeof(GPUHit) * buffer.hitScoringInfo.fSlotCounter, cudaMemcpyDefault,
+                                        cudaStreamForHitCopy));
+      COPCORE_CUDA_CHECK(cudaLaunchHostFunc(
+          cudaStreamForHitCopy,
+          [](void *arg) { static_cast<BufferHandle *>(arg)->state = BufferHandle::State::NeedHostProcessing; }, &buffer));
+    }
+  }
+  
+  std::shared_ptr<const std::vector<GPUHit>> GetNextHitsVector(unsigned int threadId)
+  {
+    assert(threadId < fHitQueues.size());
+    std::shared_lock lock{fProcessingHitsMutex};
+
+    if (fHitQueues[threadId].empty())
+      return nullptr;
+    else {
+      auto ret = fHitQueues[threadId].front();
+      fHitQueues[threadId].pop_front();
+      return ret;
+    }
+  }
+};
+
+// Implement Cuda-dependent functionality from PerEventScoring
+
+void PerEventScoring::ClearGPU(cudaStream_t cudaStream)
+{
+  COPCORE_CUDA_CHECK(cudaMemsetAsync(fScoring_dev, 0, sizeof(GlobalCounters), cudaStream));
+  COPCORE_CUDA_CHECK(cudaStreamSynchronize(cudaStream));
+}
+
+void PerEventScoring::CopyToHost(cudaStream_t cudaStream)
+{
+  const auto oldPointer = fScoring_dev;
+  COPCORE_CUDA_CHECK(
+      cudaMemcpyAsync(&fGlobalCounters, fScoring_dev, sizeof(GlobalCounters), cudaMemcpyDeviceToHost, cudaStream));
+  COPCORE_CUDA_CHECK(cudaStreamSynchronize(cudaStream));
+  assert(oldPointer == fScoring_dev);
+  (void)oldPointer;
+}
+
+// struct PerEventScoring {
+//   GlobalCounters fGlobalCounters;
+//   PerEventScoring *const fScoring_dev;
+
+//   PerEventScoring(PerEventScoring *gpuScoring) : fScoring_dev{gpuScoring} { ClearGPU(); }
+//   PerEventScoring(PerEventScoring &&other) = default;
+//   ~PerEventScoring()                       = default;
+
+//   /// @brief Copy hits to host for a single event
+//   void CopyToHost(cudaStream_t cudaStream = 0)
+//   {
+//     const auto oldPointer = fScoring_dev;
+//     COPCORE_CUDA_CHECK(
+//         cudaMemcpyAsync(&fGlobalCounters, fScoring_dev, sizeof(GlobalCounters), cudaMemcpyDeviceToHost, cudaStream));
+//     COPCORE_CUDA_CHECK(cudaStreamSynchronize(cudaStream));
+//     assert(oldPointer == fScoring_dev);
+//     (void)oldPointer;
+//   }
+
+//   /// @brief Clear hits on device to reuse for next event
+//   void ClearGPU(cudaStream_t cudaStream = 0)
+//   {
+//     COPCORE_CUDA_CHECK(cudaMemsetAsync(fScoring_dev, 0, sizeof(GlobalCounters), cudaStream));
+//     COPCORE_CUDA_CHECK(cudaStreamSynchronize(cudaStream));
+//   }
+
+//   /// @brief Print scoring info
+//   void Print() { fGlobalCounters.Print(); };
+// };
```

Comment:
Could you add a comment on this? What is the state, why is it commented out but kept?